Decorator declaration inside a function causes the typescript language server to crash

### 🔎 Search Terms

decorator language server crash

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (5.9.3 and 5.4.5), and I reviewed the FAQ for entries about "decorator"

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwM4zBgAUMAhlAObAwBccxAdMxZXvWSgJ4DaAugEo4AXgB8cTlwA0cNDkIg6cAMIAbMnjwBZGgAsIAEwAiwOVDIxoy+aBhCA3gCg4LuFBpIouZOiw43ROpowAC2wCgwOjD6BkwsVOwS3A7OrmlyKARwKGRhInDRmHiMGQRQSBjQjDl5ANRwAESMDXD1xBkKMNW5wBJ4+DBQmCiUQvUNxC1trP0aA0MjY40CDQDcqWku7jCeuIQEjARExDXAMhpc6AxCYnBOmw8BO14SAO5kmPDkVDQlZKqqUi6IoyZiMGYCdaPOAAX0hG1h6xhjkcaHUmjgAGUIGE1Bp+vdXKVBhVLFBiGAkAAjVSYNABMgGHCqLhwMD0AAKZGoDmRCIAAkcwAiLldKBBLNiwlzqMRPKp6GVhqM7gi0oLCMLoWR3p8CsDimBGOLLHKoKp4WlkciXEA

### 💻 Code

```ts
import { Page, test } from "@playwright/test";

export function step(target: (...args: any[]) => any, context: ClassMethodDecoratorContext) {
    return function replacementMethod(...args: any) {
        const name = this.constructor.name + "." + (context.name as string) + "(" + (args as string) + ")";
        return test.step(name, async () => {
            return await target.call(this, ...args);
        });
    };
}

class SomeClass {
    constructor(public readonly p: Page) {}

    @step
    async gotoSomePage(url: string) {
        @step
        await this.p.goto(url);
    }
}
```


### 🙁 Actual behavior

when (by accident) a decorator method was added into a function (line 17 / line 15 in the TS Playground), the typescript language server just gives up and exits with SIGTERM.


### 🙂 Expected behavior

mark the second decorator as wrong, or at least have the language-server exit SIGTERM with a more meaningful explanation into the Logfiles. Have it print out, in which location the language server failed. tsc just print "Error: Debug Failure. False expression." 

In my Repo, I had 10 files changed and I was looking for 4 hours to find the issue. 

### Additional information about the issue

I uploaded this to a repository here: https://github.com/devilofhell/decorator-crashes-ts-language-service/blob/main/page.ts

---

Here is a picture of the VS Code log:
<img width="1808" height="318" alt="Image" src="https://github.com/user-attachments/assets/1a190eae-30b6-493e-9ecd-01643f333444" />

Note: this only happens after you start VS Code, or open a new file, since only that triggers the language server to build up its content. VS Code just gives up after 5 retries.

---

here is the log file from /home/r975115/.vscode-server/data/logs/20251022T110530/exthost7/vscode.typescript-language-features/tsserver-log-cYXMY8/tsserver.log

[tsserver.log](https://github.com/user-attachments/files/23051094/tsserver.log)

---

if executing tsc, it says nothing meaningful. no information why or where it crashed.
```ts
/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:1284
    throw e;
    ^

Error: Debug Failure. False expression.
    at parseVariableDeclarationList (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33762:15)
    at parseVariableStatement (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33788:29)
    at parseDeclarationWorker (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33602:16)
    at parseDeclaration (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33584:14)
    at parseStatement (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33537:16)
    at parseListElement (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:30115:12)
    at parseList (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:30100:19)
    at parseBlock (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:32993:26)
    at parseFunctionBlock (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33024:19)
    at parseFunctionBlockOrSemicolon (/home/r975115/Repositories/decorator-crashes-ts-language-service/node_modules/typescript/lib/_tsc.js:33667:12)
```

This is fixed in TS7.
```
a.ts:1:1 - error TS1206: Decorators are not valid here.

1 @dec
  ~
```
Given how unlikely this mistake is, I don't think this warrants trying to fix in the 5.9/6 codebase, so we'll call this "fixed in a future release"

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -7505,7 +7505,11 @@ namespace Parser {
             case SyntaxKind.LetKeyword:
             case SyntaxKind.ConstKeyword:
             case SyntaxKind.UsingKeyword:
+                return parseVariableStatement(pos, hasJSDoc, modifiersIn);
             case SyntaxKind.AwaitKeyword:
+                if (!isAwaitUsingDeclaration()) {
+                    break;
+                }
                 return parseVariableStatement(pos, hasJSDoc, modifiersIn);
             case SyntaxKind.FunctionKeyword:
                 return parseFunctionDeclaration(pos, hasJSDoc, modifiersIn);
@@ -7534,17 +7538,16 @@ namespace Parser {
                     default:
                         return parseExportDeclaration(pos, hasJSDoc, modifiersIn);
                 }
-            default:
-                if (modifiersIn) {
-                    // We reached this point because we encountered decorators and/or modifiers and assumed a declaration
-                    // would follow. For recovery and error reporting purposes, return an incomplete declaration.
-                    const missing = createMissingNode<MissingDeclaration>(SyntaxKind.MissingDeclaration, /*reportAtCurrentPosition*/ true, Diagnostics.Declaration_expected);
-                    setTextRangePos(missing, pos);
-                    (missing as Mutable<MissingDeclaration>).modifiers = modifiersIn;
-                    return missing;
-                }
-                return undefined!; // TODO: GH#18217
         }
+        if (modifiersIn) {
+            // We reached this point because we encountered decorators and/or modifiers and assumed a declaration
+            // would follow. For recovery and error reporting purposes, return an incomplete declaration.
+            const missing = createMissingNode<MissingDeclaration>(SyntaxKind.MissingDeclaration, /*reportAtCurrentPosition*/ true, Diagnostics.Declaration_expected);
+            setTextRangePos(missing, pos);
+            (missing as Mutable<MissingDeclaration>).modifiers = modifiersIn;
+            return missing;
+        }
+        return undefined!; // TODO: GH#18217
     }
 
     function nextTokenIsStringLiteral() {

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -7677,7 +7677,9 @@ namespace Parser {
                 flags |= NodeFlags.Using;
                 break;
             case SyntaxKind.AwaitKeyword:
-                Debug.assert(isAwaitUsingDeclaration());
+                if (!isAwaitUsingDeclaration()) {
+                    break;
+                }
                 flags |= NodeFlags.AwaitUsing;
                 nextToken();
                 break;

--- a/tests/baselines/reference/decoratorOnAwait.errors.txt
+++ b/tests/baselines/reference/decoratorOnAwait.errors.txt
@@ -1,11 +1,11 @@
-decoratorOnAwait.ts(4,7): error TS1134: Variable declaration expected.
+decoratorOnAwait.ts(3,5): error TS1146: Declaration expected.
 
 
 ==== decoratorOnAwait.ts (1 errors) ====
     declare function dec<T>(target: T): T;
     
     @dec
+        
+!!! error TS1146: Declaration expected.
     await 1
-          ~
-!!! error TS1134: Variable declaration expected.
     

--- a/tests/baselines/reference/decoratorOnAwait.errors.txt
+++ b/tests/baselines/reference/decoratorOnAwait.errors.txt
@@ -1,0 +1,11 @@
+decoratorOnAwait.ts(4,7): error TS1134: Variable declaration expected.
+
+
+==== decoratorOnAwait.ts (1 errors) ====
+    declare function dec<T>(target: T): T;
+    
+    @dec
+    await 1
+          ~
+!!! error TS1134: Variable declaration expected.
+    

--- a/tests/baselines/reference/decoratorOnAwait.js
+++ b/tests/baselines/reference/decoratorOnAwait.js
@@ -1,0 +1,12 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnAwait.ts] ////
+
+//// [decoratorOnAwait.ts]
+declare function dec<T>(target: T): T;
+
+@dec
+await 1
+
+
+//// [decoratorOnAwait.js]
+var ;
+1;

--- a/tests/baselines/reference/decoratorOnAwait.js
+++ b/tests/baselines/reference/decoratorOnAwait.js
@@ -8,5 +8,4 @@ await 1
 
 
 //// [decoratorOnAwait.js]
-var ;
-1;
+await 1;

--- a/tests/baselines/reference/decoratorOnAwait.symbols
+++ b/tests/baselines/reference/decoratorOnAwait.symbols
@@ -1,0 +1,15 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnAwait.ts] ////
+
+=== decoratorOnAwait.ts ===
+declare function dec<T>(target: T): T;
+>dec : Symbol(dec, Decl(decoratorOnAwait.ts, 0, 0))
+>T : Symbol(T, Decl(decoratorOnAwait.ts, 0, 21))
+>target : Symbol(target, Decl(decoratorOnAwait.ts, 0, 24))
+>T : Symbol(T, Decl(decoratorOnAwait.ts, 0, 21))
+>T : Symbol(T, Decl(decoratorOnAwait.ts, 0, 21))
+
+@dec
+>dec : Symbol(dec, Decl(decoratorOnAwait.ts, 0, 0))
+
+await 1
+

--- a/tests/baselines/reference/decoratorOnAwait.types
+++ b/tests/baselines/reference/decoratorOnAwait.types
@@ -1,0 +1,17 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnAwait.ts] ////
+
+=== decoratorOnAwait.ts ===
+declare function dec<T>(target: T): T;
+>dec : <T>(target: T) => T
+>    : ^ ^^      ^^ ^^^^^ 
+>target : T
+>       : ^
+
+@dec
+>dec : <T>(target: T) => T
+>    : ^ ^^      ^^ ^^^^^ 
+
+await 1
+>1 : 1
+>  : ^
+

--- a/tests/baselines/reference/decoratorOnAwait.types
+++ b/tests/baselines/reference/decoratorOnAwait.types
@@ -12,6 +12,8 @@ declare function dec<T>(target: T): T;
 >    : ^ ^^      ^^ ^^^^^ 
 
 await 1
+>await 1 : 1
+>        : ^
 >1 : 1
 >  : ^
 

--- a/tests/baselines/reference/decoratorOnUsing.errors.txt
+++ b/tests/baselines/reference/decoratorOnUsing.errors.txt
@@ -1,0 +1,14 @@
+decoratorOnUsing.ts(4,7): error TS1134: Variable declaration expected.
+
+
+==== decoratorOnUsing.ts (1 errors) ====
+    declare function dec<T>(target: T): T;
+    
+    @dec
+    using 1
+          ~
+!!! error TS1134: Variable declaration expected.
+    
+    @dec
+    using x
+    

--- a/tests/baselines/reference/decoratorOnUsing.js
+++ b/tests/baselines/reference/decoratorOnUsing.js
@@ -1,0 +1,16 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnUsing.ts] ////
+
+//// [decoratorOnUsing.ts]
+declare function dec<T>(target: T): T;
+
+@dec
+using 1
+
+@dec
+using x
+
+
+//// [decoratorOnUsing.js]
+using ;
+1;
+using x;

--- a/tests/baselines/reference/decoratorOnUsing.symbols
+++ b/tests/baselines/reference/decoratorOnUsing.symbols
@@ -1,0 +1,21 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnUsing.ts] ////
+
+=== decoratorOnUsing.ts ===
+declare function dec<T>(target: T): T;
+>dec : Symbol(dec, Decl(decoratorOnUsing.ts, 0, 0))
+>T : Symbol(T, Decl(decoratorOnUsing.ts, 0, 21))
+>target : Symbol(target, Decl(decoratorOnUsing.ts, 0, 24))
+>T : Symbol(T, Decl(decoratorOnUsing.ts, 0, 21))
+>T : Symbol(T, Decl(decoratorOnUsing.ts, 0, 21))
+
+@dec
+>dec : Symbol(dec, Decl(decoratorOnUsing.ts, 0, 0))
+
+using 1
+
+@dec
+>dec : Symbol(dec, Decl(decoratorOnUsing.ts, 0, 0))
+
+using x
+>x : Symbol(x, Decl(decoratorOnUsing.ts, 6, 5))
+

--- a/tests/baselines/reference/decoratorOnUsing.types
+++ b/tests/baselines/reference/decoratorOnUsing.types
@@ -1,0 +1,25 @@
+//// [tests/cases/conformance/decorators/invalid/decoratorOnUsing.ts] ////
+
+=== decoratorOnUsing.ts ===
+declare function dec<T>(target: T): T;
+>dec : <T>(target: T) => T
+>    : ^ ^^      ^^ ^^^^^ 
+>target : T
+>       : ^
+
+@dec
+>dec : <T>(target: T) => T
+>    : ^ ^^      ^^ ^^^^^ 
+
+using 1
+>1 : 1
+>  : ^
+
+@dec
+>dec : <T>(target: T) => T
+>    : ^ ^^      ^^ ^^^^^ 
+
+using x
+>x : any
+>  : ^^^
+

--- a/tests/cases/conformance/decorators/invalid/decoratorOnAwait.ts
+++ b/tests/cases/conformance/decorators/invalid/decoratorOnAwait.ts
@@ -1,0 +1,4 @@
+declare function dec<T>(target: T): T;
+
+@dec
+await 1

--- a/tests/cases/conformance/decorators/invalid/decoratorOnUsing.ts
+++ b/tests/cases/conformance/decorators/invalid/decoratorOnUsing.ts
@@ -1,0 +1,8 @@
+// @target: esnext
+declare function dec<T>(target: T): T;
+
+@dec
+using 1
+
+@dec
+using x

